### ISSUE TITLE
6900xt not detected on linux

### ISSUE THREAD

**Issue by cdgriffith** [rigaya/VCEEnc#33]
I know linux is early experimental, just adding info to try and help! Currently not able to use on linux due to error: 

```
vceencc --check-hw
VCEEncC (x64) 6.14 (r929) by rigaya, Aug 11 2021 22:52:09 (gcc 9.3.0/Linux)
AMF Version 1.4.21
reader: raw, avs, vpy, avhw [H.264/AVC, H.265/HEVC, MPEG2, VP9, VC-1]
WARNING: radv is not a conformant vulkan implementation, testing use only.
Error (clGetDeviceIDs): device not found.
device #0: Failed to find gpu device.
Selected device #-1 not found
VCE unavailable.
```

If I try specifying 0, get the same error, if I specify 1 it changes to: 

```
$ vceencc --check-hw 1
VCEEncC (x64) 6.14 (r929) by rigaya, Aug 11 2021 22:52:09 (gcc 9.3.0/Linux)
AMF Version 1.4.21
reader: raw, avs, vpy, avhw [H.264/AVC, H.265/HEVC, MPEG2, VP9, VC-1]
WARNING: radv is not a conformant vulkan implementation, testing use only.
Error (clGetDeviceIDs): device not found.
device #0: Failed to find gpu device.
Currently default device is always used when using vulkan!: selected device = 1
```

I [custom compiled FFmpeg](https://github.com/markus-perl/ffmpeg-build-script/issues/98) with AMF headers `1.4.21` and it does work that way, so the card does support it. 

```
ffmpeg -i in.mp4 -c:v h264_amf out.mp4

...

Stream mapping:
  Stream #0:0 -> #0:0 (hevc (native) -> h264 (h264_amf))
  Stream #0:1 -> #0:1 (aac (native) -> aac (native))
```

Specs: 

* Ubuntu 20.04.2 LTS
* 6900XT 
* Driver package: `amdgpu-pro-21.30-1290604-ubuntu-20.04`

**Comment by rigaya** [comment]
I haven't been able to reproduce this issue (might because I don't have RX6xxx series), but it seems like the app is failing to find device by OpenCL.

I have improved GPU detection by OpenCL in [VCEEnc 6.15](https://github.com/rigaya/VCEEnc/releases/tag/6.15), but not sure if it works...

Would you please provide result of ```clinfo``` if you still run into error with [VCEEnc 6.15](https://github.com/rigaya/VCEEnc/releases/tag/6.15)?

**Comment by cdgriffith** [comment]
tl;dr: need to install `ocl-icd-opencl-dev`, and there are now encoding issues.

I installed open source drivers for some games, so I reinstalled the pro drivers (with rocr and legacy OpenCL support) and was getting a segmentation fault: 

```
vceencc --check-hw
VCEEncC (x64) 6.15 (r941) by rigaya, Sep  6 2021 12:41:50 (gcc 9.3.0/Linux)
AMF Version 1.4.21
reader: raw, avs, vpy, avhw [H.264/AVC, H.265/HEVC, MPEG2, VP9, VC-1]
WARNING: radv is not a conformant vulkan implementation, testing use only.
Failed to load OpenCL.
Segmentation fault (core dumped)
```


Here is clinfo:

```
Number of platforms                               1
  Platform Name                                   AMD Accelerated Parallel Processing
  Platform Vendor                                 Advanced Micro Devices, Inc.
  Platform Version                                OpenCL 2.0 AMD-APP (3314.0)
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_icd cl_amd_event_callback 
  Platform Extensions function suffix             AMD

  Platform Name                                   AMD Accelerated Parallel Processing
Number of devices                                 1
  Device Name                                     gfx1030
  Device Vendor                                   Advanced Micro Devices, Inc.
  Device Vendor ID                                0x1002
  Device Version                                  OpenCL 2.0 
  Driver Version                                  3314.0 (HSA1.1,LC)
  Device OpenCL C Version                         OpenCL C 2.0 
  Device Type                                     GPU
  Device Board Name (AMD)                         Device 73bf
  Device Topology (AMD)                           PCI-E, 03:00.0
  Device Profile                                  FULL_PROFILE
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Max compute units                               40
  SIMD per compute unit (AMD)                     4
  SIMD width (AMD)                                32
  SIMD instruction width (AMD)                    1
  Max clock frequency                             2660MHz
  Graphics IP (AMD)                               10.3
  Device Partition                                (core)
    Max number of sub-devices                     40
    Supported partition types                     None
    Supported affinity domains                    (n/a)
  Max work item dimensions                        3
  Max work item sizes                             1024x1024x1024
  Max work group size                             256
  Preferred work group size (AMD)                 256
  Max work group size (AMD)                       1024
  Preferred work group size multiple              32
  Wavefront width (AMD)                           32
  Preferred / native vector sizes                 
    char                                                 4 / 4       
    short                                                2 / 2       
    int                                                  1 / 1       
    long                                                 1 / 1       
    half                                                 1 / 1        (cl_khr_fp16)
    float                                                1 / 1       
    double                                               1 / 1        (cl_khr_fp64)
  Half-precision Floating-point support           (cl_khr_fp16)
    Denormals                                     No
    Infinity and NANs                             No
    Round to nearest                              No
    Round to zero                                 No
    Round to infinity                             No
    IEEE754-2008 fused multiply-add               No
    Support is emulated in software               No
  Single-precision Floating-point support         (core)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
  Address bits                                    64, Little-Endian
  Global memory size                              17163091968 (15.98GiB)
  Global free memory (AMD)                        16760832 (15.98GiB)
  Global memory channels (AMD)                    8
  Global memory banks per channel (AMD)           4
  Global memory bank width (AMD)                  256 bytes
  Error Correction support                        No
  Max memory allocation                           14588628168 (13.59GiB)
  Unified memory for Host and Device              No
  Shared Virtual Memory (SVM) capabilities        (core)
    Coarse-grained buffer sharing                 Yes
    Fine-grained buffer sharing                   Yes
    Fine-grained system sharing                   No
    Atomics                                       No
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       1024 bits (128 bytes)
  Preferred alignment for atomics                 
    SVM                                           0 bytes
    Global                                        0 bytes
    Local                                         0 bytes
  Max size for global variable                    14588628168 (13.59GiB)
  Preferred total size of global vars             17163091968 (15.98GiB)
  Global Memory cache type                        Read/Write
  Global Memory cache size                        16384 (16KiB)
  Global Memory cache line size                   64 bytes
  Image support                                   Yes
    Max number of samplers per kernel             29631
    Max size for 1D images from buffer            134217728 pixels
    Max 1D or 2D image array size                 8192 images
    Base address alignment for 2D image buffers   256 bytes
    Pitch alignment for 2D image buffers          256 pixels
    Max 2D image size                             16384x16384 pixels
    Max 3D image size                             16384x16384x8192 pixels
    Max number of read image args                 128
    Max number of write image args                8
    Max number of read/write image args           64
  Max number of pipe args                         16
  Max active pipe reservations                    16
  Max pipe packet size                            1703726280 (1.587GiB)
  Local memory type                               Local
  Local memory size                               65536 (64KiB)
  Local memory syze per CU (AMD)                  65536 (64KiB)
  Local memory banks (AMD)                        32
  Max number of constant args                     8
  Max constant buffer size                        14588628168 (13.59GiB)
  Preferred constant buffer size (AMD)            16384 (16KiB)
  Max size of kernel argument                     1024
  Queue properties (on host)                      
    Out-of-order execution                        No
    Profiling                                     Yes
  Queue properties (on device)                    
    Out-of-order execution                        Yes
    Profiling                                     Yes
    Preferred size                                262144 (256KiB)
    Max size                                      8388608 (8MiB)
  Max queues on device                            1
  Max events on device                            1024
  Prefer user sync for interop                    Yes
  Number of P2P devices (AMD)                     0
  P2P devices (AMD)                               <printDeviceInfo:147: get number of CL_DEVICE_P2P_DEVICES_AMD : error -30>
  Profiling timer resolution                      1ns
  Profiling timer offset since Epoch (AMD)        0ns (Wed Dec 31 18:00:00 1969)
  Execution capabilities                          
    Run OpenCL kernels                            Yes
    Run native kernels                            No
    Thread trace supported (AMD)                  No
    Number of async queues (AMD)                  8
    Max real-time compute queues (AMD)            8
    Max real-time compute units (AMD)             40
  printf() buffer size                            4194304 (4MiB)
  Built-in kernels                                (n/a)
  Device Extensions                               cl_khr_fp64 cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_int64_base_atomics cl_khr_int64_extended_atomics cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_fp16 cl_khr_gl_sharing cl_amd_device_attribute_query cl_amd_media_ops cl_amd_media_ops2 cl_khr_image2d_from_buffer cl_khr_subgroups cl_khr_depth_images cl_amd_copy_buffer_p2p cl_amd_assembly_program 

NULL platform behavior
  clGetPlatformInfo(NULL, CL_PLATFORM_NAME, ...)  No platform
  clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...)   No platform
  clCreateContext(NULL, ...) [default]            No platform
  clCreateContext(NULL, ...) [other]              Success [AMD]
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_DEFAULT)  Success (1)
    Platform Name                                 AMD Accelerated Parallel Processing
    Device Name                                   gfx1030
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  Success (1)
    Platform Name                                 AMD Accelerated Parallel Processing
    Device Name                                   gfx1030
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  Success (1)
    Platform Name                                 AMD Accelerated Parallel Processing
    Device Name                                   gfx1030
```

I did a strace to see the issue: 

```
mmap(NULL, 1576960, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fb5a4129000
brk(0x5569d376f000)                     = 0x5569d376f000
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 8
fstat(8, {st_mode=S_IFREG|0644, st_size=113314, ...}) = 0
mmap(NULL, 113314, PROT_READ, MAP_PRIVATE, 8, 0) = 0x7fb5a410d000
close(8)                                = 0
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libOpenCL.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/x86_64-linux-gnu/libOpenCL.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/lib/libOpenCL.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/libOpenCL.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
munmap(0x7fb5a410d000, 113314)          = 0
write(2, "\33[31mFailed to load OpenCL.\n\33[39"..., 33Failed to load OpenCL.
) = 33
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=NULL} ---
+++ killed by SIGSEGV (core dumped) +++
Segmentation fault (core dumped)
```

As it couldn't find the `libOpenCL.so` library at all did a quick google and saw the ubuntu package for it is `ocl-icd-opencl-dev` , probably good to add as a requirement for `deb`. 

Tried encoding: 

```
# vceencc -i artist.mp4 --vbr 5000 -o new_art.mp4
--------------------------------------------------------------------------------
new_art.mp4
--------------------------------------------------------------------------------
WARNING: radv is not a conformant vulkan implementation, testing use only.
GPU #0 (Device 73bf (40 CU) @ 2660 MHz (3314.0 (HSA1.1,LC))) does not support H.264/AVC High profile (max supported: Main).

Finished with error in VCEEncC.
```

For some reason on linux the default `profile` is set to `High` not `Baseline` as the help says. Manually changed that.

```
# vceencc -i artist.mp4 --vbr 5000 --profile Main -o new_art.mp4
--------------------------------------------------------------------------------
new_art.mp4
--------------------------------------------------------------------------------
WARNING: radv is not a conformant vulkan implementation, testing use only.
Failed to init decoder: AMF_FAIL
Finished with error in VCEEncC.
```

Hardware decoder seems to not work at all, changed to software decoder.  Finally got an encode! 

```
# vceencc --avsw -i artist.mp4 --vbr 5000 --profile Main -o new_art.mp4
--------------------------------------------------------------------------------
new_art.mp4
--------------------------------------------------------------------------------
WARNING: radv is not a conformant vulkan implementation, testing use only.
H.264/AVC reference frames should be in range of 0 - 0 (2 specified).
VCEEncC (x64) 6.15 (r941) by rigaya, Sep  6 2021 12:41:50 (gcc 9.3.0/Linux)
OS:            Ubuntu 20.04.3 LTS (5.11.0-27-generic) x64
CPU:           Intel Core i9-9900K @ 3.60GHz (8C/16T)
GPU:           Device 73bf (40 CU) @ 2660 MHz (3314.0 (HSA1.1,LC)), AMF Runtime 1.4.22 / SDK 1.4.21
Input Info:    avsw: hevc(yv12(10bit))->nv12 [AVX2], 1920x1080, 30/1 fps
Vpp Filters    copyHtoD
Output:        H.264/AVC  Main @ Level 4
               1920x1080p 0:0 30.000fps (30/1fps)
               avwriter: h264 => mp4
Quality:       balanced
VBR:           5000 kbps, Max 20000 kbps
QP:            Min: 0, Max: 51
VBV Bufsize:   25000 kbps
Bframes:       0 frames
Pre Analysis:  off
Ref frames:    0 frames
LTR frames:    0 frames
Motion Est:    Q-pel
Slices:        1
GOP Len:       300 frames
Others:        deblock 
                                                                               
encoded 264 frames, 89.98 fps, 5182.25 kbps, 5.44 MB
encode time 0:00:02, CPULoad: 0.0%
frame type IDR   1
frame type I     1,  total size  0.15 MB
frame type P   263,  total size  5.29 MB
```

And, it looks all screwy? 

![vlcsnap-2021-09-06-09h15m47s243](https://user-images.githubusercontent.com/3275435/132231397-3c13ba3d-3851-43b7-933e-930f1fe1cf3c.png)


Tried a different file and started getting a lot of errors and final result did not play: 

```
# vceencc --avsw -i cat.mp4 --vbr 5000 --profile Main -o new_cat.mp4            
--------------------------------------------------------------------------------
new_cat.mp4
--------------------------------------------------------------------------------
WARNING: radv is not a conformant vulkan implementation, testing use only.
H.264/AVC reference frames should be in range of 0 - 0 (2 specified).
VCEEncC (x64) 6.15 (r941) by rigaya, Sep  6 2021 12:41:50 (gcc 9.3.0/Linux)
OS:            Ubuntu 20.04.3 LTS (5.11.0-27-generic) x64
CPU:           Intel Core i9-9900K @ 3.60GHz (8C/16T)
GPU:           Device 73bf (40 CU) @ 2660 MHz (3314.0 (HSA1.1,LC)), AMF Runtime 1.4.22 / SDK 1.4.21
Input Info:    avsw: hevc(yv12(10bit))->nv12 [AVX2], 3840x2160, 30/1 fps
Vpp Filters    copyHtoD
Output:        H.264/AVC  Main @ Level 5.1
               3840x2160p 0:0 30.000fps (30/1fps)
               avwriter: h264 => mp4
Quality:       balanced
VBR:           5000 kbps, Max 30000 kbps
QP:            Min: 0, Max: 51
VBV Bufsize:   240000 kbps
Bframes:       0 frames
Pre Analysis:  off
Ref frames:    0 frames
LTR frames:    0 frames
Motion Est:    Q-pel
Slices:        1
GOP Len:       300 frames
Others:        deblock 
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 59000
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 148730                                                                                                                                                                            
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 37663                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 113006                                                                                                                                                                            
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 23321                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 98828                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 17886                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 92980                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 16490                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 85136                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 12827                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 81728                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 15381                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 86525                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 17018                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 84342                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 18460                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 87159                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 18318                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 81390                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 14116                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 77180                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 13148                                                                                                                                                                             
[h264 @ 0x5608e6b45dc0] missing picture in access unit with size 72011    
```

Let me know of any commands / tests I can try to help out. Thanks so much!

**Comment by rigaya** [comment]
Thank you for pointing out the ```libOpenCL.so``` issue, I added ```ocl-icd-opencl-dev``` to the dependency.

I've tried fixing the artifact caused in the encoding result, [VCEEnc 6.16](https://github.com/rigaya/VCEEnc/releases/tag/6.16) might be able to avoid it. Also default has been changed from H.264 High profile from Main profile on Linux, so you need to specify profile manually.

**Comment by cdgriffith** [comment]
It seems to be working great! The video with the artifacts looks perfect now. The other video I tested with FFmpeg and it also had errors when using h264_amf so probably something hardware side. 

I have added support for VCEEncC in [FastFlix](https://github.com/cdgriffith/FastFlix) and is working on Linux and Windows, thank you!
